boards/qemu_rv32_virt/src/io.rs

```diff
@@ -0,0 +1,60 @@
+use core::cell::Cell;
+use core::fmt::Write;
+use core::panic::PanicInfo;
+use core::str;
+
+use kernel::debug;
+use kernel::debug::IoWrite;
+use rv32i;
+
+use crate::CHIP;
+use crate::PROCESSES;
+
+struct Writer {}
+
+static mut WRITER: Writer = Writer {};
+
+impl Write for Writer {
+    fn write_str(&mut self, s: &str) -> ::core::fmt::Result {
+        self.write(s.as_bytes());
+        Ok(())
+    }
+}
+
+impl IoWrite for Writer {
+    fn write(&mut self, buf: &[u8]) {
+        let uart = qemu_rv32_virt_chip::uart::Uart16550::new(qemu_rv32_virt_chip::uart::UART0_BASE);
+        uart.transmit_sync(buf);
+    }
+}
+
+// No LEDs on the qemu-system-riscv32 virt machine "board", this is a
+// dummy type to satisfy the type constraints on [`debug::panic`].
+struct DummyLed(Cell<bool>);
+impl kernel::hil::led::Led for DummyLed {
+    fn init(&self) {
+        self.0.set(false);
+    }
+    fn on(&self) {
+        self.0.set(true);
+    }
+    fn off(&self) {
+        self.0.set(false);
+    }
+    fn toggle(&self) {
+        self.0.set(!self.0.get());
+    }
+    fn read(&self) -> bool {
+        self.0.get()
+    }
+}
+
+/// Panic handler.
+#[cfg(not(test))]
+#[no_mangle]
+#[panic_handler]
+pub unsafe extern "C" fn panic_fmt(pi: &PanicInfo) -> ! {
+    let writer = &mut WRITER;
+
+    debug::panic::<DummyLed, _, _>(&mut [], writer, pi, &rv32i::support::nop, &PROCESSES, &CHIP)
```

Comment:
You should just do what SweRVolf does here and not require the dummy LEDs: https://github.com/tock/tock/blob/master/boards/swervolf/src/io.rs#L40

Reply:
I'm not sure about this. On the one hand, the SweRVolf implementation already differs from the `kernel::panic` implementation, at least in non-substantial ways, for instance by not flushing the debug buffer (if I'm not mistaken). Further, I'd rather not have this deviate further from what the kernel does. Maybe it's best to have a `panic` and a `panic_returning` where one does not produce the infinite loop blinking?

Reply:
Opened #2534 about this. Let's see what comes out of that.

Reply:
Given that #2534 is merged I'll throw the `DummyLed` out. Thanks for the hint!